File: iRHSCBook/RHSCBookingDetailViewController.m
//
//  RHSCBookingDetailViewController.m
//  iRHSCBook
//
//  Created by Bruce Hunter on 2014-07-09.
//  Copyright (c) 2014 Richmond Hill Squash Club. All rights reserved.
//

#import "RHSCBookingDetailViewController.h"
#import <MessageUI/MessageUI.h>
#import "RHSCTabBarController.h"
#import "RHSCCourtTime.h"
#import "RHSCMemberList.h"
#import "RHSCMember.h"

@interface RHSCBookingDetailViewController ()

@property (nonatomic, strong) RHSCMember *player1;
@property (nonatomic, strong) RHSCMember *player2;
@property (nonatomic, strong) RHSCMember *player3;
@property (nonatomic, strong) RHSCMember *player4;
@property (nonatomic,strong) UIAlertView *successAlert;
@property (nonatomic,strong) UIAlertView *errorAlert;

@end

@implementation RHSCBookingDetailViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    // obtain players
    RHSCTabBarController *tbc = (RHSCTabBarController *) self.tabBarController;
    
    self.user = tbc.currentUser;
    
    NSDateFormatter* dtFormatter = [[NSDateFormatter alloc] init];
    [dtFormatter setDateFormat:@"EEE, MMM d - h:mm a"];
    
    self.courtAndTime.text = [NSString stringWithFormat:@"%@ for %@",self.booking.court,[dtFormatter stringFromDate:self.booking.courtTime]];
    self.eventLabel.text = [NSString stringWithFormat:@"%@ match with:",self.booking.event];
    self.player1 = [self findPlayer:self.booking.players[@"player1_id"] in:tbc.memberList];
    if (self.player1) {
        self.player1Label.text = [NSString stringWithFormat:@"%@ %@",self.player1.firstName,self.player1.lastName];
    } else {
        self.player1Label.text = @"";
    }
    self.player2 = [self findPlayer:self.booking.players[@"player2_id"] in:tbc.memberList];
    if (self.player2) {
        self.player2Label.text = [NSString stringWithFormat:@"%@ %@",self.player2.firstName,self.player2.lastName];
    } else {
        self.player2Label.text = @"";
    }
    if ([self.booking.court isEqualToString:@"Court 5"]) {
        self.player3 = [self findPlayer:self.booking.players[@"player3_id"] in:tbc.memberList];
        if (self.player3) {
            self.player3Label.text = [NSString stringWithFormat:@"%@ %@",self.player3.firstName,self.player3.lastName];
        } else {
            self.player3Label.text = @"";
        }
        self.player4 = [self findPlayer:self.booking.players[@"player4_id"] in:tbc.memberList];
        if (self.player4) {
            self.player4Label.text = [NSString stringWithFormat:@"%@ %@",self.player4.firstName,self.player4.lastName];
        } else {
            self.player4Label.text = @"";
        }
        
    }
}

- (RHSCMember *)findPlayer:(NSString *)playerId in:(RHSCMemberList *)memList
{
    for (RHSCMember* mem in memList.memberList)
    {
        if([mem.name isEqualToString:playerId]){
            return mem;
        }
    }
    return nil;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)cancelBooking:(id)sender {
    RHSCTabBarController *tbc = (RHSCTabBarController *)self.tabBarController;
    NSString * fetchURL = nil;
    if ([self.booking.court isEqualToString:@"Court 5"]) {
        fetchURL = [NSString stringWithFormat:@"Reserve/IOSCancelBookingJSON.php?b_id=%@&player1=%@&player2=%@&player3=%@&player4=%@&uid=%@&channel=%@",[self.booking bookingId],
                    tbc.currentUser.data.name,
                    [[self.booking players] objectForKey:@"player2_id"],
                    [[self.booking players] objectForKey:@"player3_id"],
                    [[self.booking players] objectForKey:@"player4_id"],
                    tbc.currentUser.data.name,@"iPhone"
                    ];
    } else {
        fetchURL = [NSString stringWithFormat:@"Reserve/IOSCancelBookingJSON.php?b_id=%@&player1=%@&player2=%@&player3=%@&player4=%@&uid=%@&channel=%@",[self.booking bookingId],
                          tbc.currentUser.data.name,
                          [[self.booking players] objectForKey:@"player2_id"],@"",@"",
                          tbc.currentUser.data.name,@"iPhone"
                          ];
    }
    
    NSLog(@"fetch URL = %@",fetchURL);
    NSURL *target = [[NSURL alloc] initWithString:fetchURL relativeToURL:tbc.server];
    NSURLRequest *request = [NSURLRequest requestWithURL:[target absoluteURL]
                                             cachePolicy:NSURLRequestReloadIgnoringLocalAndRemoteCacheData
                                         timeoutInterval:30.0];
    // Get the data
    NSURLResponse *response;
    NSError *error = nil;
    NSData *data = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:&error];
    if (error == nil) {
        NSDictionary *jsonDictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
        if ([jsonDictionary objectForKey:@"error"] == nil) {
            
            // Get an array of dictionaries with the key "locations"
            // NSArray *array = [jsonDictionary objectForKey:@"user"];
            NSLog(@"%@",jsonDictionary);
            self.successAlert = [[UIAlertView alloc] initWithTitle:@"Success"
                                                      message:@"Booking successfully cancelled. Notices will be sent to all players"
                                                     delegate:self
                                            cancelButtonTitle:@"OK"
                                            otherButtonTitles:nil];
            [self.successAlert show];
        } else {
            self.errorAlert = [[UIAlertView alloc] initWithTitle:@"Error"
                                                    message:[jsonDictionary objectForKey:@"error"]
                                                   delegate:self
                                          cancelButtonTitle:@"OK"
                                          otherButtonTitles:nil];
            [self.errorAlert show];
        }
    } else {
        self.errorAlert = [[UIAlertView alloc] initWithTitle:@"Error"
                                                message:[error localizedDescription]
                                               delegate:self
                                      cancelButtonTitle:@"OK"
                                      otherButtonTitles:nil];
        [self.errorAlert show];
    }
}

@synthesize delegate;

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex
{
    if (alertView == self.successAlert) {
        [delegate refreshTable];
        [self.navigationController popViewControllerAnimated:YES];
    }
}

- (IBAction)emailPlayers:(id)sender {
    if ([MFMailComposeViewController canSendMail])
    {
        RHSCTabBarController *tbc = (RHSCTabBarController *)self.tabBarController;
        RHSCMemberList *ml = tbc.memberList;
        // Email Subject
        NSString *emailTitle = @"";
        // Email Content
        NSString *messageBody = @"";
        // To address
        
        NSArray *toRecipents = [NSArray arrayWithObjects:[self.booking.players objectForKey:@"player1_id"],[self.booking.players objectForKey:@"player2_id"],[self.booking.players objectForKey:@"player3_id"],[self.booking.players objectForKey:@"player4_id"],nil];
        NSMutableArray *emailAddresses = [[NSMutableArray alloc] init];
        for (NSString *playerId in toRecipents) {
            for (RHSCMember *mem in ml.memberList) {
                if ([mem.name isEqualToString:playerId]) {
                    if (mem.email != nil) {
                        if (![mem.email isEqualToString:@"NULL"]) {
                            if (![mem.name isEqualToString:tbc.currentUser.data.name]) {
                                [emailAddresses addObject:mem.email];
                            }
                        }
                    }
                    break;
                }
            }
        }
        
        MFMailComposeViewController *mc = [[MFMailComposeViewController alloc] init];
        mc.mailComposeDelegate = self;
        [mc setSubject:emailTitle];
        [mc setMessageBody:messageBody isHTML:NO];
        [mc setToRecipients:emailAddresses];
        
        // Present mail view controller on screen
        [self presentViewController:mc animated:YES completion:NULL];
    } else {
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Cannot send email"
                                                        message:@"Cannot email from this device"
                                                       delegate:nil
                                              cancelButtonTitle:@"OK"
                                              otherButtonTitles:nil];
        [alert show];
    }
}

- (void) mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
    switch (result)
    {
        case MFMailComposeResultCancelled:
            NSLog(@"Mail cancelled");
            break;
        case MFMailComposeResultSaved:
            NSLog(@"Mail saved");
            break;
        case MFMailComposeResultSent:
            NSLog(@"Mail sent");
            break;
        case MFMailComposeResultFailed:
            NSLog(@"Mail sent failure: %@", [error localizedDescription]);
            break;
        default:
            break;
    }
    
    // Close the Mail Interface
    [self dismissViewControllerAnimated:YES completion:NULL];
}

- (IBAction)smsPlayers:(id)sender {
    RHSCTabBarController *tbc = (RHSCTabBarController *)self.tabBarController;
    RHSCMemberList *ml = tbc.memberList;
	MFMessageComposeViewController *controller = [[MFMessageComposeViewController alloc] init];
	if([MFMessageComposeViewController canSendText])
	{
        NSArray *toRecipents = [NSArray arrayWithObjects:[self.booking.players objectForKey:@"player1_id"],[self.booking.players objectForKey:@"player2_id"],[self.booking.players objectForKey:@"player3_id"],[self.booking.players objectForKey:@"player4_id"],nil];
        NSMutableArray *phoneNumbers = [[NSMutableArray alloc] init];
        for (NSString *playerId in toRecipents) {
            for (RHSCMember *mem in ml.memberList) {
                if ([mem.name isEqualToString:playerId]) {
                    if (mem.phone1 != nil) {
                        if (![mem.phone1 isEqualToString:@"NULL"]) {
                            if (![mem.name isEqualToString:tbc.currentUser.data.name]) {
                                NSString *cleanedString = [[[mem phone1] componentsSeparatedByCharactersInSet:[[NSCharacterSet characterSetWithCharactersInString:@"0123456789-+()"] invertedSet]] componentsJoinedByString:@""];
                                [phoneNumbers addObject:cleanedString];
                            }
                        }
                    }
                    break;
                }
            }
        }
		controller.body = @"";
		controller.recipients = phoneNumbers;
		controller.messageComposeDelegate = self;
        [self presentViewController:controller animated:YES completion:NULL];
	}
}

- (void)messageComposeViewController:(MFMessageComposeViewController *)controller didFinishWithResult:(MessageComposeResult)result
{
	switch (result) {
		case MessageComposeResultCancelled:
			NSLog(@"Cancelled");
			break;
		case MessageComposeResultFailed: {
            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Cannot send SMS"
                                                            message:@"Cannot message from this device"
                                                           delegate:nil
                                                  cancelButtonTitle:@"OK"
                                                  otherButtonTitles:nil];
            [alert show];
			break;
        }
		case MessageComposeResultSent:
			break;
		default:
			break;
	}
    [self dismissViewControllerAnimated:YES completion:NULL];
}


/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: iRHSCBook/RHSCBookingDetailViewController.swift
//
//  RHSCBookingDetailViewController.swift
//  iRHSCBook
//
//  Created by Bruce Hunter on 2015-10-28.
//  Copyright © 2015 Bruce Hunter. All rights reserved.
//

import Foundation
import MessageUI

protocol cancelBookingProtocol {

    func refreshTable()
}

class RHSCBookingDetailViewController : UIViewController,MFMailComposeViewControllerDelegate,MFMessageComposeViewControllerDelegate,UIAlertViewDelegate {

    @IBOutlet weak var courtAndTime : UILabel? = nil
    @IBOutlet weak var eventLabel : UILabel? = nil
    @IBOutlet weak var player1Label : UILabel? = nil
    @IBOutlet weak var player2Label : UILabel? = nil
    @IBOutlet weak var player3Label : UILabel? = nil
    @IBOutlet weak var player4Label : UILabel? = nil

    var user : RHSCUser? = nil
    var booking : RHSCCourtTime? = nil
    
    var delegate : cancelBookingProtocol? = nil
    
    var player1 : RHSCMember? = nil
    var player2 : RHSCMember? = nil
    var player3 : RHSCMember? = nil
    var player4 : RHSCMember? = nil
    var successAlert : UIAlertController? = nil
    var errorAlert : UIAlertController? = nil

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        // obtain players
        
//        self.tableView.accessibilityIdentifier = "MyBookingDetail"
        
        let tbc = tabBarController as! RHSCTabBarController
    
        user = tbc.currentUser
    
        let dtFormatter = DateFormatter()
        dtFormatter.dateFormat = "EEE, MMM d - h:mm a"
    
        courtAndTime!.text = String.init(format: "%@ for %@", arguments: [(booking?.court)!,dtFormatter.string(from: (booking?.courtTime)! as Date)])

        eventLabel!.text = String.init(format: "%@ match between:", arguments: [(booking?.event)!])
        
        player1 = booking?.players[1]
        if (player1 != nil) {
            player1Label!.text = player1!.fullName
        } else {
            player1Label!.text = ""
        }
        
        player2 = booking?.players[2]
        if (player2 != nil) {
            player2Label!.text = player2!.fullName
        } else {
            player2Label!.text = ""
        }
        
        if (booking!.court == "Court 5") {
            player3 = booking?.players[3]
            if (player3Label != nil) {
                player3Label!.text = player3!.fullName
            } else {
                player3Label!.text = ""
            }
            player4 = booking?.players[4]
            if (player4Label != nil) {
                player4Label!.text = player4!.fullName
            } else {
                player4Label!.text = ""
            }
        }
        self.view.backgroundColor = UIColor.bookedBlue()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        booking!.unlock(fromView: self)
    }
    
    func findPlayer(_ playerId : String, inList memList: RHSCMemberList) -> RHSCMember?{
        for mem in memList.memberList {
            if (mem.name == playerId){
                return mem;
            }
        }
        return nil;
    }

    @IBAction func cancelBooking(_ sender : AnyObject?) {
        booking!.cancel(fromView: self)
        booking!.unlock(fromView: self)
    }
    
    @IBAction func emailPlayers(_ sender: AnyObject) {
        if MFMailComposeViewController.canSendMail() {
            let tbc = tabBarController as! RHSCTabBarController
            let ml = tbc.memberList
            // Email Subject
            let emailTitle = ""
            // Email Content
            let messageBody = ""
            // To address
            let toRecipents = [booking!.players[1]!.name!,
                booking!.players[2]!.name!,
                booking!.players[3]!.name!,
                booking!.players[4]!.name!]
            var emailAddresses = Array<String>()
            for playerId in toRecipents {
                for mem in ml!.memberList {
                    if (mem.name == playerId) {
                        if (mem.email != nil) {
                            if (mem.email != "NULL") {
                                if (mem.name != tbc.currentUser!.name) {
                                    emailAddresses.append(mem.email!)
                                }
                            }
                        }
                        break;
                    }
                }
            }
    
            let mc = MFMailComposeViewController()
            mc.mailComposeDelegate = self;
            mc.setSubject(emailTitle)
            mc.setMessageBody(messageBody, isHTML: false)
            mc.setToRecipients(emailAddresses)
    
            // Present mail view controller on screen
            self.present(mc, animated: true, completion: nil)
        } else {
            self.errorAlert = UIAlertController(title: "Error",
                message: "Cannot email from this device", preferredStyle: .alert)
            self.present(self.errorAlert!, animated: true, completion: nil)
            let delay = 2.0 * Double(NSEC_PER_SEC)
            let time = DispatchTime.now() + Double(Int64(delay)) / Double(NSEC_PER_SEC)
            DispatchQueue.main.asyncAfter(deadline: time, execute: {
                self.errorAlert!.dismiss(animated: true, completion: nil)
            })
        }
    }
    
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        switch (result) {
        case MFMailComposeResult.cancelled:
            //            NSLog(@"Mail cancelled");
            break;
        case MFMailComposeResult.saved:
            //            NSLog(@"Mail saved");
            break;
        case MFMailComposeResult.sent:
            //            NSLog(@"Mail sent");
            break;
        case MFMailComposeResult.failed:
            //            NSLog(@"Mail sent failure: %@", [error localizedDescription]);
            break;
        default:
            break;
        }
    
        // Close the Mail Interface
        self.dismiss(animated: true, completion: nil)
    }
    
    @IBAction func smsPlayers(_ sender: AnyObject) {
        let tbc = tabBarController as! RHSCTabBarController
        let ml = tbc.memberList
        let controller = MFMessageComposeViewController()
        if MFMessageComposeViewController.canSendText() {
            let toRecipents = [booking!.players[1]!.name!,
                booking!.players[2]!.name!,
                booking!.players[3]!.name!,
                booking!.players[4]!.name!]
            var phoneNumbers = Array<String>()
            for playerId in toRecipents {
                for mem in ml!.memberList {
                    if (mem.name == playerId) {
                        if (mem.email != nil) {
                            if (mem.email != "NULL") {
                                if (mem.name != tbc.currentUser!.name) {
                                    let cleanedString = mem.phone1!
                                        .components(separatedBy: CharacterSet(charactersIn: "0123456789-+()")
                                            .inverted).joined(separator: "")
                                    phoneNumbers.append(cleanedString)
                                }
                            }
                        }
                        break;
                    }
                }
            }
            controller.body = ""
            controller.recipients = phoneNumbers
            controller.messageComposeDelegate = self
            self.present(controller, animated: true, completion: nil)
        }
    }
    
    func messageComposeViewController(_ controller: MFMessageComposeViewController, didFinishWith result: MessageComposeResult) {
        switch (result) {
        case MessageComposeResult.cancelled:
            //			NSLog(@"Cancelled");
            break;
        case MessageComposeResult.failed:
            self.errorAlert = UIAlertController(title: "Error",
                message: "Cannot message from this device", preferredStyle: .alert)
            self.present(self.errorAlert!, animated: true, completion: nil)
            let delay = 2.0 * Double(NSEC_PER_SEC)
            let time = DispatchTime.now() + Double(Int64(delay)) / Double(NSEC_PER_SEC)
            DispatchQueue.main.asyncAfter(deadline: time, execute: {
                self.errorAlert!.dismiss(animated: true, completion: nil)
            })
            break;
        case MessageComposeResult.sent:
            break;
        default:
            break;
        }
        self.dismiss(animated: true, completion: nil)
    }
    
    
}
